src/tls.c

```diff
@@ -4045,174 +4105,1901 @@ int TLSX_UseQSHScheme(TLSX** extensions, word16 name, byte* pKey, word16 pkeySz,
+    #ifdef HAVE_CURVE25519
+        case WOLFSSL_ECC_X25519:
+            break;
+    #endif
+    #ifdef HAVE_X448
+        case WOLFSSL_ECC_X448:
+            break;
+    #endif
+        default:
+            return 0;
+    }
+
+    return 1;
+}
+
+static int TLSX_KeyShare_SetSupported(WOLFSSL* ssl)
+{
+    int            ret;
+    TLSX*          extension;
+    EllipticCurve* curve = NULL;
+
+    /* Use SupportedGroup's order. */
+    extension = TLSX_Find(ssl->extensions, TLSX_SUPPORTED_GROUPS);
+    if (extension != NULL)
+        curve = (EllipticCurve*)extension->data;
+    for (; curve != NULL; curve = curve->next) {
+        if (TLSX_KeyShare_IsSupported(curve->name) &&
+                !TLSX_KeyShare_Find(ssl, curve->name)) {
+            break;
+        }
+    }
+    if (curve == NULL)
+        return BAD_KEY_SHARE_DATA;
+
+    /* Delete the old key share data list. */
+    extension = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
+    if (extension != NULL) {
+        TLSX_KeyShare_FreeAll(extension->data, ssl->heap);
+        extension->data = NULL;
+    }
+
+    /* Add in the chosen group. */
+    ret = TLSX_KeyShare_Use(ssl, curve->name, 0, NULL, NULL);
+    if (ret != 0)
+        return ret;
+
+    /* Set extension to be in reponse. */
+    extension = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
+    extension->resp = 1;
+
+    return KEY_SHARE_ERROR;
+}
+
+/* Establish the secret based on the key shares received from the client.
+ *
+ * ssl  The SSL/TLS object.
+ * returns 0 on success and other values indicate failure.
+ */
+int TLSX_KeyShare_Establish(WOLFSSL *ssl)
+{
+    int            ret;
+    TLSX*          extension;
+    KeyShareEntry* clientKSE = NULL;
+    KeyShareEntry* serverKSE;
+    KeyShareEntry* list = NULL;
+    byte*          ke;
+    word16         keLen;
+
+    /* Find the KeyShare extension if it exists. */
+    extension = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
+    if (extension != NULL)
+        list = (KeyShareEntry*)extension->data;
+
+    /* TODO: [TLS13] Server's preference and sending back SupportedGroups */
+    /* Use client's preference. */
+    for (clientKSE = list; clientKSE != NULL; clientKSE = clientKSE->next) {
+        /* Check consistency now - extensions in any order. */
+        if (!TLSX_SupportedGroups_Find(ssl, clientKSE->group))
+            return BAD_KEY_SHARE_DATA;
+
+        /* Check if server supports group. */
+        if (TLSX_KeyShare_IsSupported(clientKSE->group))
+            break;
+    }
+    /* No supported group found - send HelloRetryRequest. */
+    if (clientKSE == NULL)
+        return TLSX_KeyShare_SetSupported(ssl);
+
+    list = NULL;
+    /* Generate a new key pair. */
+    ret = TLSX_KeyShare_New(&list, clientKSE->group, ssl->heap, &serverKSE);
+    if (ret != 0)
+        return ret;
+    ret = TLSX_KeyShare_GenKey(ssl, serverKSE);
+    if (ret != 0)
+        return ret;
+
+    /* Move private key to client entry. */
+    clientKSE->key = serverKSE->key;
```

Comment:
I think this is leaking the client _ecc_key_. Valgrind is reporting
```
==12644==    by 0x418E9D: wolfSSL_Malloc (memory.c:94)
==12644==    by 0x42424A: TLSX_KeyShare_GenEccKey (tls.c:4425)
==12644==    by 0x4243FF: TLSX_KeyShare_GenKey (tls.c:4481)
==12644==    by 0x425086: TLSX_KeyShare_Use (tls.c:5005)
==12644==    by 0x42520D: TLSX_KeyShare_SetSupported (tls.c:5104)
==12644==    by 0x42530E: TLSX_KeyShare_Establish (tls.c:5148)
==12644==    by 0x45C935: VerifyServerSuite (internal.c:18897)
==12644==    by 0x45CA3C: MatchSuite (internal.c:18927)
==12644==    by 0x429D9D: DoTls13ClientHello (tls13.c:2445)
==12644==    by 0x42D143: DoTls13HandShakeMsgType (tls13.c:4688)
==12644==    by 0x42D576: DoTls13HandShakeMsg (tls13.c:4817)
```
and this line looked like it might be losing something. (I added an `XFREE` of the clientKSE key before the assignment and valgrind accepted it.)